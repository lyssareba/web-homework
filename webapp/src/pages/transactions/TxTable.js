/* eslint-disable camelcase, react/jsx-sort-props */
import React, { useState } from 'react'
import { arrayOf, string, bool, number, shape } from 'prop-types'
import { useMutation, useQuery } from '@apollo/client'
import GetUsers from '../../gql/users.gql'

import TableCx from '../../components/table'
import { TableContext } from '../../components/table/TableContext'

import AddTransaction from '../../gql/addTransaction.gql'
import UpdateTransaction from '../../gql/updateTransaction.gql'
import DeleteTransaction from '../../gql/deleteTransaction.gql'

export const tableHeaderKeys = [
  { id: 'id', label: 'ID', readOnly: true, placeholder: 'Auto Generated', type: 'id' },
  { id: 'user', label: 'User', placeholder: 'employee1', type: 'name', typeFulfillmentKeys: ['firstName', 'lastName'], inputType: 'dropdown' },
  { id: 'description', label: 'Description', placeholder: 'groceries', type: 'text' },
  { id: 'merchant_id', label: 'Merchant ID', placeholder: 'walmart', type: 'text' },
  { id: 'amount', label: 'Amount', placeholder: '-100 or +100', type: 'currency' }
]

const createData = (data) => data.map(({ id, user_id, description, merchant_id, debit, amount, user }) => ({
  id,
  user,
  user_id,
  description,
  merchant_id,
  amount,
  debit,
  isEditMode: false
}))

const TxTable = ({ data }) => {
  const [addTransaction] = useMutation(AddTransaction)
  const [updateTransaction] = useMutation(UpdateTransaction)
  const [deleteTransaction] = useMutation(DeleteTransaction)
  const { data: usersData } = useQuery(GetUsers)
  const formattedData = createData(data)
  const [rows, setRows] = useState(formattedData)
  const [previous, setPrevious] = useState({})

  const testPrefix = 'transactions'
  const makeDataTestId = (transactionId, fieldName) => `${testPrefix}-${transactionId}-${fieldName}`
  const txObject = {
    id: null,
    user: null,
    user_id: null,
    description: null,
    merchant_id: null,
    amount: null,
    debit: null,
    isEditMode: null
  }

  const onSave = async (id, rows, setRows, previous) => {
    if (id === '') {
      const newTransaction = rows.find(row => row.id === '')
      newTransaction.credit = /^\+/.test(newTransaction.amount) || false
      newTransaction.debit = /^-/.test(newTransaction.amount) || false
      newTransaction.amount = parseFloat(newTransaction.amount.substring(1))
      delete newTransaction.isEditMode
      delete newTransaction.id
      delete newTransaction.user
      try {
        const { data } = await addTransaction({
          variables: {
            ...newTransaction
          }
        })
        newTransaction.user = usersData?.users.find(usr => usr.id === newTransaction.user_id)
        const newRows = rows.map(row => {
          if (row.id === undefined) {
            const formattedRow = Object.assign(txObject, row)
            formattedRow.id = data.addTransaction.id
            formattedRow.isEditMode = false
            return { ...formattedRow }
          }
          return row
        })
        setRows(newRows)
      } catch (err) {
        console.log(err)
      }
    } else {
      const updatedTransaction = rows.find(row => row.id === id)

      if (updatedTransaction.amount !== previous[id].amount) {
        updatedTransaction.credit = /^\+/.test(updatedTransaction.amount) || false
        updatedTransaction.debit = /^-/.test(updatedTransaction.amount) || false
        updatedTransaction.amount = parseFloat(updatedTransaction.amount.substring(1))
      }
      delete updatedTransaction.isEditMode
      delete updatedTransaction.user

      try {
        await updateTransaction({
          variables: {
            id: updatedTransaction.id,
            user_id: updatedTransaction.user_id,
            description: updatedTransaction.description,
            merchant_id: updatedTransaction.merchant_id,
            debit: updatedTransaction.debit,
            credit: updatedTransaction.credit,
            amount: updatedTransaction.amount
          }
        })

        updatedTransaction.user = usersData?.users.find(usr => usr.id === updatedTransaction.user_id)
        setRows(state => {
          return rows.map(row => {
            if (row.id === id) {
              delete row.credit
              const formattedRow = Object.assign(txObject, row)
              return formattedRow
            }
            return row
          })
        })
        setPrevious(state => {
          delete state[id]
          return state
        })
      } catch (err) {
        console.log(err)
      }
    }
  }

  const onTransactionDelete = async id => {
    try {
      await deleteTransaction({
        variables: { id }
      })
    } catch (err) {
      console.log(err)
    }
  }

  const onAddTransactionClick = (rows, setRows) => {
    const newRows = [...rows]
    const emptyRow = {
      id: '',
      user: {},
      user_id: '',
      description: '',
      merchant_id: '',
      amount: '',
      isEditMode: true
    }
    newRows.unshift(emptyRow)
    setRows(newRows)
  }

  const formatDropdownData = () => {
    const { users } = usersData
    const formattedUsers = users.map(({ id, lastName, firstName }) => ({
      id,
      name: firstName.concat(' ', lastName)
    }))
    return formattedUsers
  }

  return (
    <TableContext.Provider
      value={{
        rows,
        setRows,
        previous,
        setPrevious,
        tableHeaderKeys,
        makeDataTestId
      }}
    >
      <TableCx.Header
        addItemText={'Add Transaction'}
        onAddClick={onAddTransactionClick}
        tableTitle={'Transactions'}
        testPrefix={testPrefix}
      />
      <TableCx
        onSave={onSave}
        onDelete={onTransactionDelete}
        inputDropdownData={{
          key: 'user',
          updateKey: 'user_id',
          data: usersData?.users?.length && formatDropdownData()
        }}
      />
    </TableContext.Provider>
  )
}

TxTable.propTypes = {
  data: arrayOf(shape({
    id: string,
    user_id: string,
    description: string,
    merchant_id: string,
    debit: bool,
    credit: bool,
    amount: number
  }))
}

export default TxTable
