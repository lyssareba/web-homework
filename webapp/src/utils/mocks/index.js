export * from './transactions-data'
export * from './users-data'