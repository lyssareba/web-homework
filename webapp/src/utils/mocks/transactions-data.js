export const transactions = [
  {
    "id": "6185992085f9270693b1dd76",
    "user_id": "61958780fdbd1a43b62d22be",
    "description": "beer",
    "merchant_id": "walmart",
    "debit": true,
    "credit": false,
    "amount": 2.49,
    "user": {
      "id": "61958780fdbd1a43b62d22be",
      "dob": "01/01/1000",
      "firstName": "The",
      "lastName": "Doctor",
      "__typename": "TxUser"
    },
    "__typename": "Transaction"
  },
  {
    "id": "6193f15d29bfd8875c649bfb",
    "user_id": "619587bdfdbd1a43b62d22bf",
    "description": "office supplies",
    "merchant_id": "staples",
    "debit": false,
    "credit": true,
    "amount": 100,
    "user": {
      "id": "619587bdfdbd1a43b62d22bf",
      "dob": "6/25/1982",
      "firstName": "Billie",
      "lastName": "Pieper",
      "__typename": "TxUser"
    },
    "__typename": "Transaction"
  },
  {
    "id": "619463389d0b3bb69eba7379",
    "user_id": "61946c571c2a20c8f553d76b",
    "description": "baking supplies",
    "merchant_id": "walmart",
    "debit": true,
    "credit": false,
    "amount": 14.99,
    "user": {
      "id": "61946c571c2a20c8f553d76b",
      "dob": "01/25/1000",
      "firstName": "Clara",
      "lastName": "Oswald",
      "__typename": "TxUser"
    },
    "__typename": "Transaction"
  }
]
