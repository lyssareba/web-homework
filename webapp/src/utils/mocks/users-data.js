export const users = [
  {
    "id": "61946c571c2a20c8f553d76b",
    "dob": "01/25/1000",
    "lastName": "Oswald",
    "firstName": "Clara",
    "transactions": [
      {
        "id": "619463389d0b3bb69eba7379",
        "description": "baking supplies",
        "merchant_id": "walmart",
        "amount": 14.99,
        "debit": true,
        "credit": null,
        "__typename": "Transaction"
      }
    ],
    "__typename": "User"
  },
  {
    "id": "6195869efdbd1a43b62d22bd",
    "dob": "01/01/1900",
    "lastName": "Harkness",
    "firstName": "Jack",
    "transactions": [],
    "__typename": "User"
  },
  {
    "id": "61958780fdbd1a43b62d22be",
    "dob": "01/01/1000",
    "lastName": "Doctor",
    "firstName": "The",
    "transactions": [
      {
        "id": "6185992085f9270693b1dd76",
        "description": "beer",
        "merchant_id": "walmart",
        "amount": 2.49,
        "debit": true,
        "credit": null,
        "__typename": "Transaction"
      }
    ],
    "__typename": "User"
  },
  {
    "id": "619587bdfdbd1a43b62d22bf",
    "dob": "6/25/1982",
    "lastName": "Pieper",
    "firstName": "Billie",
    "transactions": [
      {
        "id": "6193f15d29bfd8875c649bfb",
        "description": "office supplies",
        "merchant_id": "staples",
        "amount": 100,
        "debit": false,
        "credit": true,
        "__typename": "Transaction"
      }
    ],
    "__typename": "User"
  }
]
