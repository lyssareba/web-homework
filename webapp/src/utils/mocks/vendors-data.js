export const vendors = [
  {
    "id": "61a175d0388a4073ca8b0f85",
    "name": "Staples"
  },
  {
    "id": "61a175e6388a4073ca8b0f86",
    "name": "Walmart"
  },
  {
    "id": "61a175f3388a4073ca8b0f87",
    "name": "Pet Smart"
  },
  {
    "id": "61a175f9388a4073ca8b0f88",
    "name": "Starbucks"
  },
  {
    "id": "61a175ff388a4073ca8b0f89",
    "name": "Lowes"
  }
]
