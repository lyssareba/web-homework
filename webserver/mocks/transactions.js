const transactions = [
  {
    "_id": "6193f15d29bfd8875c649bfb",
    "user_id": "619587bdfdbd1a43b62d22bf",
    "description": "Fish and Chips",
    "debit": true,
    "credit": null,
    "amount": 12.99,
    "vendor_id": "61a175f9388a4073ca8b0f88",
    "category": "FOOD"
  },
  {
    "_id": "619463389d0b3bb69eba7379",
    "user_id": "61946c571c2a20c8f553d76b",
    "description": "Baking Supplies",
    "debit": true,
    "credit": false,
    "amount": 12.99,
    "vendor_id": "61a175e6388a4073ca8b0f86",
    "category": "SUPPLIES"
  },
  {
    "_id": "61a037616c191f0669d0959c",
    "user_id": "6195869efdbd1a43b62d22bd",
    "description": "Dog Food",
    "debit": true,
    "credit": false,
    "amount": 59.99,
    "vendor_id": "61a175f3388a4073ca8b0f87",
    "category": "SUPPLIES"
  },
  {
    "_id": "61a0412c6c191f0669d0959d",
    "user_id": "61946c571c2a20c8f553d76b",
    "description": "Pastries",
    "debit": true,
    "credit": false,
    "amount": 49,
    "vendor_id": "61a175f9388a4073ca8b0f88",
    "category": "FOOD"
  },
  {
    "_id": "61a15f27e0c35260802cc074",
    "user_id": "61958780fdbd1a43b62d22be",
    "description": "Screw Driver",
    "debit": true,
    "credit": false,
    "amount": 1.99,
    "vendor_id": "61a175ff388a4073ca8b0f89",
    "category": "HARDWARE"
  }
]

module.exports = transactions
