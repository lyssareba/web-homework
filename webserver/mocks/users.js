const users = [
  {
    "_id": "61946c571c2a20c8f553d76b",
    "dob": "01/25/1000",
    "firstName": "Clara",
    "lastName": "Oswald"
  },
  {
    "_id": "6195869efdbd1a43b62d22bd",
    "dob": "01/01/1900",
    "firstName": "Jack",
    "lastName": "Harkness"
  },
  {
    "_id": "61958780fdbd1a43b62d22be",
    "dob": "01/01/1000",
    "firstName": "The",
    "lastName": "Doctor"
  },
  {
    "_id": "619587bdfdbd1a43b62d22bf",
    "dob": "6/25/1982",
    "firstName": "Billie",
    "lastName": "Pieper"
  }
]

module.exports = users