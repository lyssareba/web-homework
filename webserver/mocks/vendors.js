const vendors = [
  {
    "_id": "61a175d0388a4073ca8b0f85",
    "name": "Staples"
  },
  {
    "_id": "61a175e6388a4073ca8b0f86",
    "name": "Walmart"
  },
  {
    "_id": "61a175f3388a4073ca8b0f87",
    "name": "Pet Smart"
  },
  {
    "_id": "61a175f9388a4073ca8b0f88",
    "name": "Starbucks"
  },
  {
    "_id": "61a175ff388a4073ca8b0f89",
    "name": "Lowes"
  }
]

module.exports = vendors
