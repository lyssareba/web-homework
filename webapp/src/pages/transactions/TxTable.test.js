import { fireEvent, render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { MockedProvider } from '@apollo/client/testing'
import TxTable from './TxTable'
import { transactions, users } from '../../utils/mocks'
import GetUsers from '../../gql/users.gql'
import DeleteTransaction from '../../gql/deleteTransaction.gql'
import UpdateTransaction from '../../gql/updateTransaction.gql'
import AddTransaction from '../../gql/addTransaction.gql'


describe('Transactions Table', () => {
  const testTransaction = transactions.find(tx => tx.user_id === users[0].id)

  const mockGetUsers = {
    request: {
      query: GetUsers
    },
    result: {
      data: {
        users
      }
    }
  }
  
  const mockDeleteTx = {
    request: {
      query: DeleteTransaction,
      variables: { id: testTransaction.id }
    },
    result: {
      data: {
        id: null,
        __typename: 'Transaction'
      }
    }
  }

  const mockAddTx = {
    request: {
      query: AddTransaction,
      variables: {
        user_id: '61946c571c2a20c8f553d76b',
        description: 'test description',
        merchant_id: 'test merchant',
        debit: true,
        credit: false,
        amount: 2.99
      }
    },
    result: {
      data: {
        addTransaction: {
          id: 'testid',
        },
      }
    }
  }

  const mockUpdateTx = {
    request: {
      query: UpdateTransaction,
      variables: {
        id: '619463389d0b3bb69eba7379',
        user_id: '61946c571c2a20c8f553d76b',
        description: 'baking supplies',
        merchant_id: 'walmart',
        debit: true,
        credit: false,
        amount: 17.99
      }
    },
    result: {
      data: {
        updateTransaction: {
          id: null,
        },
      }
    }
  }

  const mocks = [ mockGetUsers, mockDeleteTx, mockAddTx, mockUpdateTx ]

  it('should match snapshot', () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })
  it('should show correct user data', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    const txDataRow = getByTestId(`transactions-${testTransaction.id}-table-row`)
    expect(txDataRow).toBeInTheDocument()
  })
  it('should switch to edit mode when edit icon clicked', async () => {
    const { getByTestId } = await render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )

    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txDataEditButton = getByTestId(`transactions-${testTransaction.id}-edit-button`)
    fireEvent.click(txDataEditButton)
    const txDescriptionInputField = getByTestId(`transactions-${testTransaction.id}-description-input`)
    expect(txDescriptionInputField).toBeInTheDocument()
  })
  it('should add transaction row when Add Transaction button clicked', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txAddButton = getByTestId(`transactions-add-button`)
    fireEvent.click(txAddButton)
    const newInputField = getByTestId(`transactions--description-input`)
    expect(newInputField).toBeInTheDocument()
  })
  it('should save new transaction row when Add Transaction row filled and saved', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypeName={false}>
        <TxTable data={transactions} />
      </MockedProvider>
    )
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txAddButton = getByTestId(`transactions-add-button`)
    fireEvent.click(txAddButton)

    const newDescriptionField = getByTestId(`transactions--description-input`)
    fireEvent.change(newDescriptionField, { target: { value: 'test description' } })

    const newMerchantField = getByTestId(`transactions--merchant_id-input`)
    fireEvent.change(newMerchantField, { target: { value: 'test merchant' } })

    const newUserField = getByTestId(`transactions--user-input-dropdown`)
    fireEvent.change(newUserField, { target: { value: users[0].id } })

    const newAmountField = getByTestId(`transactions--amount-input`)
    fireEvent.change(newAmountField, { target: { value: '-2.99' } })

    const saveNewTxField = getByTestId(`transactions--save-button`)
    fireEvent.click(saveNewTxField)

    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const newTxRow = getByTestId(`transactions-testid-table-row`)
    expect(newTxRow).toBeInTheDocument()
  })
  it('should delete row when delete button clicked', async () => {
    const { getByTestId, queryByTestId } = render(
      <MockedProvider mocks={mocks} addTypename={false} >
        <TxTable data={transactions} />
      </MockedProvider>
    )
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txDataDeleteButton = getByTestId(`transactions-${testTransaction.id}-delete-button`)
    fireEvent.click(txDataDeleteButton)
  
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const deletedRow = queryByTestId(`transactions-${testTransaction.id}-table-row`)
    expect(deletedRow).not.toBeInTheDocument()
  })
  it('should clear changes when transaction updated and revert button clicked', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypename={false} >
        <TxTable data={transactions} />
      </MockedProvider>
    )
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txDataEditButton = getByTestId(`transactions-${testTransaction.id}-edit-button`)
    fireEvent.click(txDataEditButton)

    const txDescriptionInputField = getByTestId(`transactions-${testTransaction.id}-description-input`)
    fireEvent.change(txDescriptionInputField, { target: { value: 'test description', name: 'description' }})

    const txDataRevertIconButton = getByTestId(`transactions-${testTransaction.id}-revert-button`)
    fireEvent.click(txDataRevertIconButton)

    const txDescriptionCell = getByTestId(`transactions-${testTransaction.id}-description-cell`)
    expect(txDescriptionCell).toHaveTextContent(testTransaction.description)
  })
  it('should save changes when transaction updated', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypename={false} >
        <TxTable data={transactions} />
      </MockedProvider>
    )
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txDataEditButton = getByTestId(`transactions-${testTransaction.id}-edit-button`)
    fireEvent.click(txDataEditButton)

    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txAmountInputField = getByTestId(`transactions-${testTransaction.id}-amount-input`)
    fireEvent.change(txAmountInputField, { target: { value: '-17.99' }})

    const txDataSaveIconButton = getByTestId(`transactions-${testTransaction.id}-save-button`)
    fireEvent.click(txDataSaveIconButton)

    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

    const txAmountCell = getByTestId(`transactions-${testTransaction.id}-amount-cell`)
    expect(txAmountCell).toHaveTextContent('17.99')
  })
})