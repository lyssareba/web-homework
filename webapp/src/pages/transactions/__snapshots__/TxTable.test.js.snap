// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Table should match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-heading-1 MuiGrid-container MuiGrid-align-content-xs-center MuiGrid-justify-content-xs-space-between"
  >
    <h1
      class="MuiTypography-root MuiTypography-h4"
    >
      Transactions
    </h1>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="transactions-add-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
        Add Transaction
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiTableContainer-root"
    >
      <table
        class="MuiTable-root MuiTable-stickyHeader"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              User
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              Description
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              Merchant ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
              scope="col"
            >
              Amount
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
            data-testid="transactions-6185992085f9270693b1dd76-table-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-selectTableCell-2"
            >
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transactions-6185992085f9270693b1dd76-edit-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transactions-6185992085f9270693b1dd76-delete-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6185992085f9270693b1dd76-id-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                93b1dd76
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6185992085f9270693b1dd76-user-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                The Doctor
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6185992085f9270693b1dd76-description-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                beer
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6185992085f9270693b1dd76-merchant_id-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                walmart
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6185992085f9270693b1dd76-amount-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="color: red;"
              >
                - $2.49
              </p>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
            data-testid="transactions-6193f15d29bfd8875c649bfb-table-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-selectTableCell-2"
            >
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transactions-6193f15d29bfd8875c649bfb-edit-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transactions-6193f15d29bfd8875c649bfb-delete-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6193f15d29bfd8875c649bfb-id-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                5c649bfb
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6193f15d29bfd8875c649bfb-user-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Billie Pieper
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6193f15d29bfd8875c649bfb-description-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                office supplies
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6193f15d29bfd8875c649bfb-merchant_id-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                staples
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-6193f15d29bfd8875c649bfb-amount-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="color: green;"
              >
                + $100.00
              </p>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
            data-testid="transactions-619463389d0b3bb69eba7379-table-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-selectTableCell-2"
            >
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transactions-619463389d0b3bb69eba7379-edit-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="transactions-619463389d0b3bb69eba7379-delete-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-619463389d0b3bb69eba7379-id-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                9eba7379
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-619463389d0b3bb69eba7379-user-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Clara Oswald
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-619463389d0b3bb69eba7379-description-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                baking supplies
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-619463389d0b3bb69eba7379-merchant_id-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                walmart
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-3 MuiTableCell-alignLeft"
              data-testid="transactions-619463389d0b3bb69eba7379-amount-cell"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="color: red;"
              >
                - $14.99
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
